File: blockchain_transactions.py

import hashlib
import time

class Transaction:
    def __init__(self, sender, receiver, amount):
        self.sender = sender
        self.receiver = receiver
        self.amount = amount

class Block:
    def __init__(self, index, timestamp, transactions, previous_hash):
        self.index = index
        self.timestamp = timestamp
        self.transactions = transactions
        self.previous_hash = previous_hash
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        data_to_hash = f"{self.index}{self.timestamp}{self.transactions}{self.previous_hash}"
        return hashlib.sha256(data_to_hash.encode()).hexdigest()

def create_block_with_transactions(previous_block, transactions):
    index = previous_block.index + 1
    timestamp = str(time.time())
    return Block(index, timestamp, transactions, previous_block.hash)

def add_block_to_chain(blockchain, block):
    blockchain.append(block)

if __name__ == "__main__":
    blockchain = [Block(0, str(time.time()), [], "0")]

    tx1 = Transaction("Andrej", "BabaYaga", 10)
    tx2 = Transaction("BabaYaga", "Dead Party", 5)

    new_block = create_block_with_transactions(blockchain[0], [tx1.__dict__, tx2.__dict__])
    add_block_to_chain(blockchain, new_block)

    for block in blockchain:
        print(f"Bloco: {block.__dict__}")

File: blockchain_transactions.go

package main

import (
    "crypto/sha256"
    "encoding/hex"
    "fmt"
    "strconv"
    "time"
)

// Estrutura de uma transação
type Transaction struct {
    Sender    string
    Receiver  string
    Amount    int
}

// Estrutura de um bloco
type Block struct {
    Index        int
    Timestamp    string
    Data         string
    PreviousHash string
    Hash         string
}

// Função para calcular o hash de um bloco
func calculateHash(block Block) string {
    record := strconv.Itoa(block.Index) + block.Timestamp + block.Data + block.PreviousHash
    hash := sha256.New()
    hash.Write([]byte(record))
    hashed := hash.Sum(nil)
    return hex.EncodeToString(hashed)
}

// Função para criar um novo bloco com transações
func createBlockWithTransaction(prevBlock Block, transactions []Transaction) Block {
    var txData string
    for _, tx := range transactions {
        txData += fmt.Sprintf("%s enviou %d moedas para %s\n", tx.Sender, tx.Amount, tx.Receiver)
    }
    newBlock := Block{
        Index:        prevBlock.Index + 1,
        Timestamp:    time.Now().String(),
        Data:         txData,
        PreviousHash: prevBlock.Hash,
        Hash:         "",
    }
    newBlock.Hash = calculateHash(newBlock)
    return newBlock
}

func main() {
    genesisBlock := Block{0, time.Now().String(), "Bloco Gênese", "", ""}
    genesisBlock.Hash = calculateHash(genesisBlock)
    Blockchain = append(Blockchain, genesisBlock)

    // Criando algumas transações
    tx1 := Transaction{"Andrej", "BabaYaga", 10}
    tx2 := Transaction{"BabaYaga", "Dead Party", 5}

    // Criando um novo bloco com transações
    newBlock := createBlockWithTransaction(genesisBlock, []Transaction{tx1, tx2})
    addBlock(newBlock)

    fmt.Println("Blockchain com transações:", Blockchain)
}
